src/test/java/lambda/part3/exercise/Mapping.java

```diff
@@ -150,22 +181,111 @@ public LazyFlatMapHelper(List<T> list, Function<T, List<R>> function) {
         // flatMap": [T1, T2] -> (T -> [T]) -> [T2]
 
         public <R2> LazyFlatMapHelper<T, R2> map(Function<R, R2> f) {
-            final Function<R, List<R2>> listFunction = rR2TorListR2(f);
-            return flatMap(listFunction);
+            Function<R, List<R2>> listFunc = f.andThen((R2 r) -> {
+                ArrayList<R2> arrList = new ArrayList<>();
+                arrList.add(r);
+                return arrList;
+            });
+            return flatMap(listFunc);
         }
 
-        // (R -> R2) -> (R -> [R2])
-        private <R2> Function<R, List<R2>> rR2TorListR2(Function<R, R2> f) {
-            throw new UnsupportedOperationException();
+        // TODO *
+        public <R2> LazyFlatMapHelper<T, R2> flatMap(Function<R, List<R2>> func) {
+            Function<T, List<R2>> listFunc = t -> new MapHelper<>(function.apply(t)).flatMap(func).getList();
+            return new LazyFlatMapHelper<>(list, listFunc);
         }
+    }
 
-        // TODO *
-        public <R2> LazyFlatMapHelper<T, R2> flatMap(Function<R, List<R2>> f) {
-            throw new UnsupportedOperationException();
+
+    interface Traversable<T> {
+        void forEach(Consumer<T> consumer);
+
+        default <R> Traversable<R> map(Function<T, R> func) {
+            Traversable<T> self = this;
+            Traversable<R> newTravers = new Traversable<R>() {
+                @Override
+                public void forEach(Consumer<R> consumer) {
+                    self.forEach(t -> consumer.accept(func.apply(t)));
+                }
+            };
+            return newTravers;
+        }
+
+        default <R> Traversable<R> flatmap(Function<T, List<R>> f) {
+            Traversable<T> self = this;
+            return new Traversable<R>() {
+                @Override
+                public void forEach(Consumer<R> consumer) {
+                    self.forEach((T t) -> f.apply(t).forEach(p -> consumer.accept(p)));

```

Comment:
What is the type of `consumer`? What is the type of `p -> consumer.accept(p)` here?